File: apps/render-ios-playground/render-ios-playground/Presentation/Controllers/RenderController.swift
import UIKit

protocol Renderer {
    var type: String { get }
    func render(component: Component) -> UIView?
}

/// Controller responsible for rendering components to UIViews
class RenderController {
    private var renderers: [String: Renderer] = [:]
    
    init(renderers: [Renderer]) {
        for renderer in renderers {
            self.renderers[renderer.type] = renderer
        }
    }
    
    func render(_ component: Component) -> UIView? {
        guard let renderer = renderers[component.type] else {
            return nil
        }
        
        guard let view = renderer.render(component: component) else {
            return nil
        }
        
        for child in component.getChildren() {
            if let subview = self.render(child) {
                view.addSubview(subview)
            }
        }
        
        return view
    }
}

File: apps/render-android-playground/app/src/main/java/com/max_mrtnv/render_android_playground/domain/services/RenderController.kt
package com.max_mrtnv.render_android_playground.domain.services

import android.view.View
import android.view.ViewGroup
import com.max_mrtnv.render_android_playground.domain.entities.Component

/**
 * Controller responsible for rendering components to Android Views
 */
class RenderController(
    private val renderers: Map<String, Renderer>
) {
    fun render(component: Component, container: ViewGroup): View? {
        val renderer = renderers[component.type] ?: return null
        val view = renderer.render(component)

        view?.let {
            // Remove from parent if already attached
            (it.parent as? ViewGroup)?.removeView(it)

            // Add to container
            container.addView(it)
        }

        return view
    }
}

/**
 * Interface for component renderers
 */
interface Renderer {
    val type: String

    fun render(component: Component): View?
}